File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    int count(ListNode* head) {
        int cnt = 0;
        while(head) {
            head = head->next;
            cnt++;
        }
        return cnt;
    }
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int cnt1 = count(headA);
        int cnt2 = count(headB);
        if(cnt1 > cnt2) {
            while(headA && cnt1 > cnt2) {
                headA = headA->next;
                cnt1--;
            }
        }
        else if(cnt2 > cnt1) {
            while(headB && cnt2 > cnt1) {
                headB = headB->next;
                cnt2--;
            }
        }
        while(headA != headB) {
            headA = headA->next;
            headB = headB->next;
        }
        return headA;
    }
};

File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        int len1 = 0, len2 = 0;
        ListNode temp1 = headA, temp2 = headB;
        while(temp1 != null) {
            len1++;
            temp1 = temp1.next;
        }
        while(temp2 != null) {
            len2++;
            temp2 = temp2.next;
        }
        temp1 = headA;
        temp2 = headB;
        while(len1 > len2) {
            len1--;
            temp1 = temp1.next;
        }
        while(len1 < len2) {
            len2--;
            temp2 = temp2.next;
        }
        while(temp1 != temp2) {
            temp1 = temp1.next;
            temp2 = temp2.next;
        }
        return temp1;
    }
}